modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotMetadataVerificationTask.java

```diff
@@ -88,25 +96,74 @@ public MetadataVerificationJob(SnapshotMetadataVerificationTaskArg arg) {
         }
 
         /** {@inheritDoc} */
-        @Override public List<SnapshotMetadata> execute() throws IgniteException {
+        @Override public List<SnapshotMetadata> execute() {
             IgniteSnapshotManager snpMgr = ignite.context().cache().context().snapshotMgr();
 
-            List<SnapshotMetadata> snpMeta = snpMgr.readSnapshotMetadatas(arg.snapshotName(), arg.snapshotPath());
+            List<SnapshotMetadata> metas = snpMgr.readSnapshotMetadatas(arg.snapshotName(), arg.snapshotPath());
+
+            for (SnapshotMetadata meta : metas)
+                checkMeta(meta, ignite.context().config().getEncryptionSpi().masterKeyDigest());
 
             if (arg.incrementIndex() > 0) {
-                List<SnapshotMetadata> metas = snpMeta.stream()
+                List<SnapshotMetadata> locNodeMetas = metas.stream()
                     .filter(m -> m.consistentId().equals(ignite.localNode().consistentId()))
                     .collect(Collectors.toList());
 
-                if (metas.size() != 1) {
-                    throw new IgniteException("Failed to find snapshot metafile [metas=" + metas +
-                        ", snpName=" + arg.snapshotName() + ", snpPath=" + arg.snapshotPath() + ']');
+                if (locNodeMetas.size() != 1) {
+                    throw new IgniteException("Failed to find single snapshot metafile for local node [locNodeId="
+                        + ignite.localNode().consistentId() + ", metas=" + metas + ", snpName=" + arg.snapshotName()
+                        + ", snpPath=" + arg.snapshotPath() + ']');
+                }
+
+                checkIncrementalSnapshots(locNodeMetas.get(0), arg);
+            }
+
+            return metas;
+        }
+
+        /** */
+        private void checkMeta(SnapshotMetadata meta, byte[] masterKeyDigest) {
+            byte[] snpMasterKeyDigest = meta.masterKeyDigest();
+
+            if (masterKeyDigest == null && snpMasterKeyDigest != null) {
+                throw new IllegalStateException("Snapshot '" + meta.snapshotName() + "' has encrypted caches " +
+                    "while encryption is disabled. To restore this snapshot, start Ignite with configured " +
+                    "encryption and the same master key.");
+            }
+
+            if (snpMasterKeyDigest != null && !Arrays.equals(snpMasterKeyDigest, masterKeyDigest)) {
+                throw new IllegalStateException("Snapshot '" + meta.snapshotName() + "' has different master " +
+                    "key digest. To restore this snapshot, start Ignite with the same master key.");
+            }
+
+            Collection<Integer> grpIds = new HashSet<>(F.isEmpty(arg.grpIds()) ? meta.cacheGroupIds() : arg.grpIds());
+
+            if (meta.hasCompressedGroups() && grpIds.stream().anyMatch(meta::isGroupWithCompression)) {
+                try {
+                    if (ignite.context().compress() == null)
```

Comment:
Do actually we need this check? I suppose that user should add ignite-compress module to classpath and this message is in the catch exception block below. Let's catch the NPE with the IgniteCheckedException together.